src/fpylll/tools/probabilistic_bkz_simulator.py

```diff
@@ -0,0 +1,215 @@
+        {"i":        1,  "r_0":   2^32.7,  "r_0/gh": 3.905694,  "rhf": 1.016064,  "/": -0.06171,  "hv/hv": 2.162941}
+        {"i":        2,  "r_0":   2^32.4,  "r_0/gh": 3.222033,  "rhf": 1.015087,  "/": -0.05777,  "hv/hv": 2.055214}
+        {"i":        3,  "r_0":   2^32.1,  "r_0/gh": 2.536317,  "rhf": 1.013873,  "/": -0.05587,  "hv/hv": 2.005962}
+
+
+    """
+
+    if not prng_seed:
+        prng_seed = FPLLL.randint(0, 2**32-1)
+
+    random.seed(prng_seed)
+
+    if isinstance(r, IntegerMatrix):
+        r = GSO.Mat(r)
+    if isinstance(r, MatGSO):
+        r.update_gso()
+        r = r.r()
+
+    n = len(r)
+
+    # code uses log2 of norms, FPLLL uses squared norms
+    r = list(map(lambda x: log(x, 2) / 2.0, r))
+
+    l = copy(r)
+    l̂ = copy(r)
+    c = [rk[-j] - sum(rk[-j:]) / j for j in range(1, 46)]
+    c += [
+        (lgamma(d / 2.0 + 1) * (1.0 / d) - log(sqrt(pi))) / log(2.0)
+        for d in range(46, param.block_size + 1)
+    ]
+
+    if param.max_loops:
+        N = param.max_loops
+    else:
+        N = n
+
+    t0 = [True for _ in range(n)]
+    for j in range(N):
+        t1 = [False for _ in range(n)]
+        for k in range(n - min(45, param.block_size)):
+            d = min(param.block_size, n - k - 1)
+            e = k + d + 1
+            tau = False
+            for kp in range(k, e):
+                tau |= t0[kp]
+            logV = sum(l[:e-1]) - sum(l̂[:k])
+            if tau:
+                X = random.expovariate(.5)
+                g = (log(X, 2) + logV) / d + c[d - 1]
+                if g < l[k]:
+                    l̂[k] = g
+                    l̂[k+1] = l[k] + log(sqrt(1-1./d), 2)
+                    γ = (l[k] + l[k+1]) - (l̂[k] + l̂[k+1])
+                    for kp in range(k+2, e):
+                        l̂[kp] = l[kp] + γ/(d-2.)
+                        t1[kp] = True
+                    tau = False
+            for idx in range(k, e-1):
+                l[idx] = l̂[idx]
+
+
+        # early termination
+        if True not in t1 or l == l̂:
+            break
+        else:
+            d = min(45, param.block_size)
+            logV = sum(l) - sum(l̂[:-d])
+
+            if param.block_size < 45:
+                tmp = sum(rk[-param.block_size :]) / param.block_size
+                rk1 = [r_ - tmp for r_ in rk[-param.block_size :]]
+            else:
+                rk1 = rk
+
+            for k, r in zip(range(n - d, n), rk1):
+                l̂[k] = logV / d + r
+                t1[kp] = True
+            l = copy(l̂)
+            t0 = copy(t1)
+
+        if param.flags & BKZ.VERBOSE:
+            r = OrderedDict()
+            r["i"] = j
+            for k, v in basis_quality(list(map(lambda x: 2.0 ** (2 * x), l))).items():
+                r[k] = v
+            print(pretty_dict(r))
+
+    l = list(map(lambda x: 2.0 ** (2 * x), l))
+    return l, j + 1
+
+
+def averaged_simulate(L, params_fplll, tries=10):
```

Comment:
We're a bit punkrock in general with `param` v `params` but I suggest to unify it in this file. I tend towards `params`

Reply:
sure